src/allmydata/testing/web.py

```diff
@@ -0,0 +1,225 @@
+import hashlib
+
+import attr
+
+from hyperlink import DecodedURL
+
+from twisted.web.resource import (
+    Resource,
+)
+from twisted.web.iweb import (
+    IBodyProducer,
+)
+from twisted.internet.defer import (
+    succeed,
+)
+
+from treq.client import (
+    HTTPClient,
+    FileBodyProducer,
+)
+from treq.testing import (
+    RequestTraversalAgent,
+)
+from zope.interface import implementer
+
+import allmydata.uri
+from allmydata.util import (
+    base32,
+)
+
+
+class _FakeTahoeRoot(Resource, object):
+    """
+    This is a sketch of how an in-memory 'fake' of a Tahoe
+    WebUI. Ultimately, this will live in Tahoe
+    """
+
+    def __init__(self, uri=None):
+        Resource.__init__(self)  # this is an old-style class :(
+        self._uri = uri
+        self.putChild(b"uri", self._uri)
+
+    def add_data(self, kind, data):
+        return self._uri.add_data(kind, data)
+
+
+@attr.s
+class _FakeCapability(object):
+    """
+    """
+    data=attr.ib()
+
+
+KNOWN_CAPABILITIES = [
+    getattr(allmydata.uri, t).BASE_STRING
+    for t in dir(allmydata.uri)
+    if hasattr(getattr(allmydata.uri, t), 'BASE_STRING')
+]
+
+
+def capability_generator(kind):
+    """
+    :param str kind: the kind of capability, like `URI:CHK`
```

Comment:
This docstring could use a heading.  It looks like the point of this function is to deterministically create an unbounded sequence of cap strings?

Is there a big advantage to making these look cryptographic instead of just using counting integers in a sufficiently wide zero-padded field?

Reply:
more like just the first thing i thought of :)

Reply:
(but also: it needs to be valid base32 .. which isn't "all numbers", but only 2-7)

Reply:
I guess we could base32-encode a padded number, but then it's also going to look like garbage too so I just left it..